File: DatabaseClient.java
import database.DatabaseOuterClass.DBRequest;
import database.DatabaseOuterClass.DBResponse;
import database.DatabaseOuterClass.OperationType;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;

public class DatabaseClient {

    public static void main(String[] args) throws IOException {
        String host = "localhost";
        int port = 9090;

        try (Socket socket = new Socket(host, port);
             InputStream in = socket.getInputStream();
             OutputStream out = socket.getOutputStream()) {

            // 1️⃣ Add records
            int[] indices = {4101, 4102, 4103, 4104, 4105, 4106};
            String[] values = {"Appen", "Ahrensburg", "Wedel", "Aumühle", "Seevetal", "Quickborn"};

            for (int i = 0; i < indices.length; i++) {
                DBRequest req = DBRequest.newBuilder()
                        .setOperation(OperationType.ADD_RECORD)
                        .setIndex(indices[i])
                        .setRecord(values[i])
                        .build();
                req.writeDelimitedTo(out);

                DBResponse resp = DBResponse.parseDelimitedFrom(in);
                System.out.println("Add " + indices[i] + ": " + resp.getSuccess());
            }

            // 2️⃣ Read records 4103 and 4107
            int[] queryIndices = {4103, 4107};
            for (int idx : queryIndices) {
                DBRequest req = DBRequest.newBuilder()
                        .setOperation(OperationType.GET_RECORD)
                        .setIndex(idx)
                        .build();
                req.writeDelimitedTo(out);

                DBResponse resp = DBResponse.parseDelimitedFrom(in);
                if (resp.getSuccess()) {
                    System.out.println("Record " + idx + ": " + resp.getRecord());
                } else {
                    System.out.println("Record " + idx + " not found (" + resp.getError() + ")");
                }
            }

            // 3️⃣ Get database size
            DBRequest sizeReq = DBRequest.newBuilder()
                    .setOperation(OperationType.GET_SIZE)
                    .build();
            sizeReq.writeDelimitedTo(out);

            DBResponse sizeResp = DBResponse.parseDelimitedFrom(in);
            System.out.println("Database size: " + sizeResp.getSize());
        }
    }
}

File: gRPC_python/DatabaseClient.py
import grpc
import Database_pb2
import Database_pb2_grpc


def run(host='localhost', port=9090):
    channel = grpc.insecure_channel(f"{host}:{port}")
    stub = Database_pb2_grpc.DatabaseStub(channel)

    records = [
        (4101, "Appen"),
        (4102, "Ahrensburg"),
        (4103, "Wedel"),
        (4104, "Aumühle"),
        (4105, "Seevetal"),
        (4106, "Quickborn"),
    ]

    print("Adding records...")
    for idx, val in records:
        req = Database_pb2.AddRecordRequest(index=idx, value=val)
        res = stub.AddRecord(req)
        print(f" AddRecord({idx}, {val!r}) -> ok={res.ok}")

    read_idx = [4103,4107]
    for idx in read_idx:
        print(f"\nReading record {idx}...")
        r = stub.GetRecord(Database_pb2.GetRecordRequest(index=idx))
        if r.found:
            print(f" GetRecord({idx}) -> '{r.value}'")
        else:
            print(f" GetRecord({idx}) -> NOT FOUND")

    print("\nCalling GetSize()...")
    s = stub.GetSize(Database_pb2.GetSizeRequest())
    print(f" GetSize() -> {s.size}")


if __name__ == '__main__':
    run()